### PR TITLE
add some more tests, use pytest instead of unittest.TestCase

### PR DESCRIPTION
**Pull request by fuhrysteve**
I added a couple tests for a couple trivial things, then converted all the tests to pytest (since it's getting used anyways). Wonder if you could take a quick look then merge - then I think I can merge your pull request without coveralls complaining too, since this should actually bump us up to 100% coverage.

One thing I noticed is that your patch won't work with marshmallow 2.3, since one of the validators (Range? I can't remember) was breaking, so I bumped up the requirement to the most recent version of marshmallow. Let me know if you guys have any issues with that.

### CODE REVIEW

File: tests/test_dump.py [hunk 1]
```diff
@@ -1,119 +1,177 @@
...
+    assert dumped['properties']['const']['minLength'] == 50
+    assert dumped['properties']['const']['maxLength'] == 50
+
+def test_length_validator_value_error():
+    class BadSchema(Schema):
+        bob = fields.Integer(validate=validate.Length(min=1, max=3))
+    schema = BadSchema(strict=True)
+    json_schema = JSONSchema()
+    with pytest.raises(ValueError):
+        json_schema.dump(schema)
```
Comment (fuhrysteve) [inline tests/test_dump.py:93]: added this test

File: tests/test_dump.py [hunk 2]
```diff
@@ -1,119 +1,177 @@
...
+        floor = fields.Integer(validate=validate.Range(min=1, max=4))
+    class SchemaWithIntRangeNoValidate(Schema):
+        floor = fields.Integer()
+    schema1 = SchemaWithStringRange(strict=True)
+    schema2 = SchemaWithNoRange(strict=True)
+    schema3 = SchemaWithIntRangeValidate(strict=True)
+    schema4 = SchemaWithIntRangeNoValidate(strict=True)
+    json_schema = JSONSchema()
+    json_schema.dump(schema1) == json_schema.dump(schema2)
+    json_schema.dump(schema3) != json_schema.dump(schema4)
```
Comment (fuhrysteve) [inline tests/test_dump.py:111]: and this one

File: tests/test_dump.py [hunk 3]
```diff
@@ -1,119 +1,177 @@
...
+        floor = fields.Integer(validate=validate.Range(max=4))
+    schema1 = SchemaMin(strict=True)
+    schema2 = SchemaNoMin(strict=True)
+    json_schema = JSONSchema()
+    dumped1 = json_schema.dump(schema1)
+    dumped2 = json_schema.dump(schema2)
+    dumped1.data['properties']['floor']['minimum'] == 1
+    dumped1.data['properties']['floor']['exclusiveMinimum'] is True
+    dumped2.data['properties']['floor']['minimum'] == 0
+    dumped2.data['properties']['floor']['exclusiveMinimum'] is False
```
Comment (fuhrysteve) [inline tests/test_dump.py:127]: and this one

**Comment by eligundry** [comment]
@fuhrysteve this all looks good. I'm curious as to why all the cool kids are switching to pytest now?

**Comment by fuhrysteve** [comment]
I think a couple reasons
- nosetests is basically unsupported now, and nose2, which is aimed at replacing it, doesn't seem quite mature (I don't remember what issues I had with nose2)
- pytest is actively developed & supported
- pytest's fixtures are a bit more useful for reusing things than trying to inject the same boilerplate `setUp` and `tearDown` in every test
- no need to memorize weird custom `assertFoo` - just use assert, because pytest does some impressive magic when there's a failure (try running a test using pytest for `assert {'foo':'bar'} == {'foo':'111'}` - the output is very useful, and you don't need `assertEqual` to do it

Good section to skim through http://pytest.org/latest/fixture.html

The "gotcha" is that everything revolves around `conftest.py` files. `conftest.py` and fixtures are a tid bit magical / not intuitive at first, but once you understand the magic, it is very useful and flexible.

I've been using pytest a lot lately, and it's just coming out a lot cleaner than trying to cram everything into `unittest.TestCase`

It's probably more useful in larger projects where you have > 100 tests.

Hope that's helpful! Lots of great tools out there.

**Comment by hjc** [comment]
@fuhrysteve - that was a super useful explanation of PyTest! We're currently using nose in our projects (I work with @eligundry), but since it's EoL'd, we've been looking for a new tool. I completely agree that `nose2`, while a good idea, isn't the mature replacement you want (I remember looking for **colored output** and finding _nothing_!). I started looking into pytest recently, and really **loved** how they did fixtures, so it's what I'm starting to lean towards. After talking to @eligundry today, I think he'll take pytest over nose2!

Our current project has about 300 tests, some are unit tests and others are full functional tests. Trying to cram them all into one `unittest.TestCase` that fits them all is... doable, but requires a LOT of backflips. I'm starting to like pytest more and more as I read it!

**Comment by fuhrysteve** [comment]
Glad to hear I'm not the only one! And yes, I was about at your point when I started to feel the pain & limits of `unittest.TestCase` when I started looking for something different, and found pytest fixtures.

I've recently also started using [factory-boy](http://factoryboy.readthedocs.io), particularly for testing code that needs an instance of a sqlalchemy model. If you use sqlalchemy or django or something similar, take a look at it and see if you might find it useful. Another thing you might find interesting, [this video](https://www.youtube.com/watch?v=RAxiiRPHS9k) has sort-of reshaped the way I think about testing web applications.

Best!

**Comment by hjc** [comment]
Awesome! Thanks for sharing!

We use Faker for our data factory library and use PeeWee for an ORM, but it looks like there's a PeeWee extension for factory-boy as well, so I'll look into it (we're like... 75% in love with Faker though).

Will definitely give that vid a shot if you recommend it so highly!

**Comment by fuhrysteve** [comment]
FactoryBoy & faker play nice together! The documentation isn't the greatest, but it works quite well:

http://factoryboy.readthedocs.io/en/latest/reference.html#faker

**Comment by hjc** [comment]
Oh nice! Thanks for the link! I'll keep that in mind as I play with factory-boy.

**Pull request merged**